### ISSUE TITLE
[no-invalid-this] triggers on class fields

### ISSUE THREAD

**Issue by PEConn** [typescript-eslint/typescript-eslint#2840]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{ 
  "no-invalid-this": "off",
  "@typescript-eslint/no-invalid-this": ["error"]
}
```

```TS
enum LogLevel { Debug, Info, Warn, Error }

/** Data class to store level and message of pending */
class PendingLog {
  constructor(public level: LogLevel, public message: string) {}
}

/**
 * A Log that wraps another Log, saving up all the log calls to be applied when flush is called.
 *
 * It doesn't currently support arguments to the log messages.
 */
export class BufferedLog implements Log {
  private pendingLogs: Array<PendingLog> = [];

  constructor(private innerLog: Log) {}

  /* eslint-disable no-invalid-this */
  debug = this.addLogFunction(LogLevel.Debug);
  info = this.addLogFunction(LogLevel.Info);
  warn = this.addLogFunction(LogLevel.Warn);
  error = this.addLogFunction(LogLevel.Error);
  /* eslint-enable no-invalid-this */

  /** Creates a function that adds a log at the given level. */
  private addLogFunction(level: LogLevel): (message: string) => void {
    return (message: string): void => {
      this.pendingLogs.push(new PendingLog(level, message));
    };
  }

  // ...
}
```

**Expected Result**

I believe my use of this between the eslint-disable comments is valid. (I could be wrong.)

**Actual Result**

Without /* eslint-disable no-invalid-this */ I get:
```
  41:11  error  Unexpected 'this'  @typescript-eslint/no-invalid-this
  42:10  error  Unexpected 'this'  @typescript-eslint/no-invalid-this
  43:10  error  Unexpected 'this'  @typescript-eslint/no-invalid-this
  44:11  error  Unexpected 'this'  @typescript-eslint/no-invalid-this
```

**Additional Info**

I previously reported this for eslint [here](https://github.com/eslint/eslint/issues/13894), and they directed me to use the typescript specific rule, which also triggered.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `^2.34.0` |
| `@typescript-eslint/parser`        | `^2.34.0` |
| `TypeScript`                       | `^3.9.7` |
| `ESLint`                           | `^6.6.0` |
| `node`                             | `v12.14.1` |

**Comment by bradzacher** [comment]
> - [x] I have updated to the latest version of the packages.

You ticked the box. But you're on v2.34

Our latest release is v4.9

Please follow the template. Its intended to save us both time.
